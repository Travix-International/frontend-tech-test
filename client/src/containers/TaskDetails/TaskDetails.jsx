import React, { useEffect } from 'react';
import { connect } from 'react-redux';

import { fetchTask } from '../../store/actions/taskActions';
import Task from '../../components/Task/Task';
import Loader from '../../components/UI/Loader/Loader';

const TaskDetails = props => {

  useEffect(() => {
    if(props.match.params.id) {
      props.fetchTask(props.match.params.id);
    }
  }, [])

  return (
    <div>
      {!props.task ? 
        <Loader /> :
        <Task task={props.task} />}
    </div>
  );
}

const mapStateToProps = state => ({
  task: state.task.task
})

export default connect(mapStateToProps, { fetchTask })(TaskDetails);
