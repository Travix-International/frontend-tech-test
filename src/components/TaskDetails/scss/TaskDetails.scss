$breakpoint-tablet: 768px;

.taskDetailsContainer {
    position: fixed;
    top: 0px;
    background-color: #ffffffb3;
    width: 100%;
    height: 100vh;
    display: flex;
    align-items: center;
    justify-content: right;
}

    .contents{
        position: fixed;
        width: 100%;
        height: 100%;
        background-color: #fff;
        box-sizing: border-box;
        box-shadow: 0 3px 8px rgba(0, 0, 0, .25);
        padding: 15px;
        display: flex;
        flex-direction: column;
        top: 0px;
        right: 0px;

        .taskInfo{
            flex-grow: 1;
        }

        .closeButton{
            float: right;
            font-size: 40px;
            color: #f82d2dcc;
        }

        .loader{
            text-align: center;
            margin-top: 200px;
            font-weight: 800;
            font-size: 26px;
        }

        .taskTitle{
            font-size: 34px;
            font-weight: 600;
            padding-bottom: 15px;
            border: none;
            display: block;
            width: 100%;
        }

        .taskDescription{
            margin-top: 15px;
            font-size: 14px;
            padding-bottom: 15px;
            max-height: 300px;
            border: none;
            display: block;
            width: 100%;
            height: 100%;
            padding-left: 4px;
            box-sizing: border-box;
            font-family: helvetica;
            line-height: 22px;
            font-size: 14px;
            color: #454545;
        }

        [contenteditable=true]:empty:before{
            content: attr(placeholder);
            display: block;
        }

        .saveButton{
            height: 45px;
            padding: 2px 25px;
            border-radius: 28px;
            border: none;
            cursor: pointer;
            font-size: 16px;
            background-color: #6eb77e;
            border-color: #6eb77e;
            color: #fff;
            font-weight: 600;
            width: 200px;
            align-self: center;
            margin-bottom: 20px;
        }
    }

@media only screen and (min-width: $breakpoint-tablet) {
    .contents{
        width: 45%;
    }
}