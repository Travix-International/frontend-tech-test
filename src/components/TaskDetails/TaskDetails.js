import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import FaClose from 'react-icons/lib/fa/times-circle';
import './scss/TaskDetails.scss';

class TaskDetails extends Component{
    constructor(props){
        super(props);
        this.state = {
            status: false,
            message: 'Loading..',
            info : {
                id: null,
                title: null,
                description: null,
            }
        }
    }

    render(){
        return (
            <div>
                <div className="taskDetailsContainer"></div>
                <form className="contents" onSubmit={this.submitChanges}>
                <div><Link to="/" ><FaClose className="closeButton" /></Link></div>
                { 
                    this.state.status
                    ? <div className="taskInfo" >
                        <input type="text" className="taskTitle" placeholder="Task title goes here.." defaultValue={this.state.info.title} />
                        <textarea className="taskDescription" placeholder="Add task description here.." defaultValue={this.state.info.description} ></textarea>
                    </div>
                    : <div className="loader">{this.state.message}</div>
                }
                
                <button type="submit" className="saveButton" id="saveButton">SAVE</button>
                </form>
            </div>
        );
    }

    submitChanges = (event) => {
        event.preventDefault();
        let taskTitle = event.target[0].value;
        let desctiption = event.target[1].value;
        this.props.update(this.state.info.id, taskTitle, desctiption);
    }

    componentDidMount() {
		fetch(`http://localhost:9001${location.pathname}`)
		  .then(res => res.json())
		  .then(
			(result) => {
			  this.setState({
                info: result.task,
                status: true
			  });
			},
			(error) => {
			  this.setState({
				message: 'Error!'
			  });
			}
		  )
	  }

}

export default TaskDetails;