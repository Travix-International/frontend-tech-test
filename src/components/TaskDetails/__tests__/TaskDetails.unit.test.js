import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TaskDetails from '../TaskDetails';

test('Fetches task details and renders properly', () => {
    fetch.mockResponses(
		[
			JSON.stringify({task: {description: "Task description", id: 0, title: "Task Title",}}),
			{ status: 200 }
		]
    );
    const updateTaskDetails = () => {};
    const wrapper = shallow(<TaskDetails update={updateTaskDetails}/>);;
    expect(toJson(wrapper)).toMatchSnapshot();
});

test('Call update method on click on save button.', () => {
    fetch.mockResponses(
		[
			JSON.stringify({task: {description: "Task description", id: 0, title: "Task Title",}}),
			{ status: 200 }
		]
    );
    const updateTaskDetails =jest.fn();
    const wrapper = shallow(<TaskDetails update={updateTaskDetails}/>);;
    wrapper.find("form").simulate("submit", {
            preventDefault: () => {},
            target: ["Task Title","Task description"]
    });
    expect(updateTaskDetails.mock.calls.length).toBe(1);
});
