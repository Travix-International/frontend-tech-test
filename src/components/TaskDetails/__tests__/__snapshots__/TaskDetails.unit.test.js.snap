// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetches task details and renders properly 1`] = `
<div>
  <div
    className="taskDetailsContainer"
  />
  <form
    className="contents"
    onSubmit={[Function]}
  >
    <div>
      <Link
        replace={false}
        to="/"
      >
        <FaTimesCircle
          className="closeButton"
        />
      </Link>
    </div>
    <div
      className="loader"
    >
      Loading..
    </div>
    <button
      className="saveButton"
      id="saveButton"
      type="submit"
    >
      SAVE
    </button>
  </form>
</div>
`;
